### ISSUE TITLE
timers in python

### ISSUE THREAD

**Issue by ZoliveGit** [jruizgit/rules#35]
Hi,

I have a problem with "timers" :

```
from durable.lang import *
with ruleset('VMC'):
	@when_all(m.Cmd == 'On')
	def start_timer(c):
		print("Start")
		c.start_timer('my_timer', 5)

	@when_all(timeout('my_timer'))
	def end_timer(c):
		print('Stop')

run_all()
```

It works only the first time I got the "Start" and 5sec later I got the "Stop", the second time I only got the "Start".

To make it work I have to change the name of the timer, for example 'my_timer' in 'my_timer1' restart the script and it works again one time.

The name of the timer is maybe save as an ID and it's work just one time ? or the timeout fonction destroy the object 'my_timer' ?

Thank's for your reply

**Comment by jruizgit** [comment]
Hi, thanks for posting the question. start_timer takes an id as third parameter, your scenario should work if you pass a different id every time you call the function. I recommend using the context state 'c.s' to keep track of the id sequence.

Sent from my phone

On Aug 24, 2016, at 8:45 AM, ZoliveGit <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi,

I have a problem with "timers" :

from durable.lang import *
with ruleset('VMC'):
    @when_all(m.Cmd == 'On')
    def start_timer(c):
        print("Start")
        c.start_timer('my_timer', 5)

    @when_all(timeout('my_timer'))
    def end_timer(c):
        print('Stop')

run_all()


It works only the first time I got the "Start" and 5sec later I got the "Stop", the second time I only got the "Start".

To make it work I have to change the name of the timer, for example 'my_timer' in 'my_timer1' restart the script and it works again one time.

The name of the timer is maybe save as an ID and it's work just one time ? or the timeout fonction destroy the object 'my_timer' ?

Thank's for your reply

-

**Comment by ZoliveGit** [comment]
Thank's for your answer,

I try this :

```
from durable.lang import *
with ruleset('VMC'):
	@when_all(m.Cmd == 'On')
	def start_timer(c):
		c.start_timer('my_timer', 5, c.m.id)
		print('my_timer %s' % c.m.id)

	@when_all(timeout('my_timer'))
	def end_timer(c):
		print('Stop')

run_all()
```

the id value com from my call script

```
url = 'http://127.0.0.1:5000/VMC/1'
payload = {'id':id_Rules, 'Cmd':Cmd}
headers = {'content-type': 'application/json'}
r = requests.post(url, data=json.dumps(payload), headers=headers)
```

I got the log "my_timer 99" (if c.m.id= 99) but never the log "Stop"... I think i miss something

Thank's for your help

**Comment by shinrairis** [comment]
As far as I know, all the events with the same sid in durable_rules require unique ids. You might need to give a different id to the timer event. 

The following example works for me.
```python
from durable.lang import *

with ruleset('VMC'):
    @when_all(m.Cmd == 'On')
    def start_timer(c):
        c.start_timer('my_timer', 5, c.m.id + 1)
        print('my_timer %s' % (c.m.id + 1)) 

    @when_all(timeout('my_timer'))
    def end_timer(c):
        print('Stop')

    @when_start
    def start(host):
        host.post('VMC', {'id': 1, 'sid': 1, 'Cmd': 'On'})

run_all()
```

**Comment by jruizgit** [comment]
Yes, thanks for posting the sample. Timer ids and message ids need to be unique.

**Comment by ZoliveGit** [comment]
Thanks for your answer it's work well.

Shinrairis you say "ll the events with the same sid in durable_rules require unique ids". Does it mean if i change the sid, id can be unique for timer ? i think i don't understand well the functioning of sid

**Comment by shinrairis** [comment]
Quote from #21
> A sid identifies an inference session. Inference context and state are keyed by sid. For instance if action A depends on Message B and Message C and Message C is observed, the fact that message C has been oseeved is stored in a HashSet keyed by sid. 
> An id uniquely identifies a Message. It ensures the idempotency of post, assert and retract operations. @jruizgit 

Please correct me if I am wrong, @jruizgit . 
From my understanding, a message id only needs to be unique in an inference session, which means that it is okay for messages with different sid to have the same id.
